core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -32,156 +32,185 @@
    */
   @Override
   public final ClientStream newStream(MethodDescriptor<?, ?> method, Metadata headers,
       CallOptions callOptions, StatsTraceContext statsTraceCtx) {
-    Supplier<ClientTransport> supplier = transportSupplier;
-    if (supplier == null) {
+    try {
+      SubchannelPicker picker = null;
+      long pickerVersion = -1;
       synchronized (lock) {
-        // Check again, since it may have changed while waiting for lock
-        supplier = transportSupplier;
-        if (supplier == null && !shutdown) {
-          if (backoffStatus != null && !callOptions.isWaitForReady()) {
-            return new FailingClientStream(backoffStatus);
+        if (!shutdown) {
+          if (lastPicker == null) {
+            return createPendingStream(method, headers, callOptions, statsTraceCtx);
+          }
+          picker = lastPicker;
+          pickerVersion = lastPickerVersion;
+        }
+      }
+      if (picker != null) {
+        while (true) {
+          PickResult pickResult = picker.pickSubchannel(callOptions.getAffinity(), headers);
+          ClientTransport transport = GrpcUtil.getTransportFromPickResult(
+              pickResult, callOptions.isWaitForReady());
+          if (transport != null) {
+            return transport.newStream(method, headers, callOptions, statsTraceCtx);
           }
-          PendingStream pendingStream = new PendingStream(method, headers, callOptions,
-              statsTraceCtx);
-          pendingStreams.add(pendingStream);
-          if (pendingStreams.size() == 1) {
-            listener.transportInUse(true);
+          // This picker's conclusion is "buffer".  If there hasn't been a newer picker set
+          // (possible race with reprocess()), we will buffer it.  Otherwise, will try with the new
+          // picker.
+          synchronized (lock) {
+            if (shutdown) {
+              break;
+            }
+            if (pickerVersion == lastPickerVersion) {
+              return createPendingStream(method, headers, callOptions, statsTraceCtx);
+            }
+            picker = lastPicker;
+            pickerVersion = lastPickerVersion;
           }
-          return pendingStream;
         }
       }
+      return new FailingClientStream(Status.UNAVAILABLE.withDescription(
+              "Channel has shutdown (reported by delayed transport)"));
+    } finally {
+      channelExecutor.drain();
     }
-    if (supplier != null) {
-      return supplier.get().newStream(method, headers, callOptions, statsTraceCtx);
-    }
-    return new FailingClientStream(Status.UNAVAILABLE.withDescription("transport shutdown"));
   }
 
   @Override
   public final ClientStream newStream(MethodDescriptor<?, ?> method, Metadata headers) {
     return newStream(method, headers, CallOptions.DEFAULT, StatsTraceContext.NOOP);
   }
 
+  // Caller must call channelExecutor.drain() outside of lock because this method may schedule
```

Comment:
This should be in javadoc format, because some tools don't automatically associate such comments with the method.